Requestify the Raw Value Accessor

CodaFi opened the pull request:
Make getRawValueExpr() return a checked value.

This entails a strange kind of request that effectively acts like
a cache warmer.  In order to properly check the raw value expression for
a single case, we actually need all the other cases for the
autoincrementing synthesis logic.  The strategy is therefore to have the
request act at the level of the parent EnumDecl and check all the values
at once.  We also cache at the level of the EnumDecl so the cache
"warms" for all enum elements simultaneously.

The request also abuses TypeResolutionStage to act as an indicator for
how much information to compute.  In the minimal case, we will return
a complete accounting of (auto-incremented) raw values.  In the maximal
case we will also check and record types and emit diagnostics.  The
minimal case is uncached to support repeated evaluation.

With this change, computing the interface type of a nominal type decl is
finally a trivial operation.

CodaFi committed:
Requestify the Raw Value Accessor

Make getRawValueExpr() return a checked value.

This entails a strange kind of request that effectively acts like
a cache warmer.  In order to properly check the raw value expression for
a single case, we actually need all the other cases for the
autoincrementing synthesis logic.  The strategy is therefore to have the
request act at the level of the parent EnumDecl and check all the values
at once.  We also cache at the level of the EnumDecl so the cache
"warms" for all enum elements simultaneously.

The request also abuses TypeResolutionStage to act as an indicator for
how much information to compute.  In the minimal case, we will return
a complete accounting of (auto-incremented) raw values.  In the maximal
case we will also check and record types and emit diagnostics.  The
minimal case is uncached to support repeated evaluation.

Note that computing the interface type of an @objc enum decl *must*
force this request.  The enum's raw values are part of the ABI, and we
should not get all the way to IRGen before discovering that we cannot
possibly lay out the enum.  In the future, we might want to consider
moving this check earlier or have IRGen tolerate broken cases but for
now we will maintain the status quo and not have IRGen emit
diagnostics.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
For the reviewers: The deleted test was checking to see that we would emit diagnostics in non-primaries in order to head off a crash later.  I suspect the actual cause was the old raw value synthesis code not being called on all validation paths.  Now that we only force this request in the decl checker, and IRGen is perfectly capable of firing off this request, there's never going to be a situation where we need to diagnose here in a non-primary.

CodaFi commented:
@swift-ci please clean smoke test

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1466,61 +1463,82 @@ computeAutomaticEnumValueKind(TypeChecker &TC, EnumDecl *ED) {
+      assert(TC && "Must have a lazy resolver set");
+      // Try to pull out the automatic enum value kind.  If that fails, bail.
+      if (!valueKind) {
+        valueKind = computeAutomaticEnumValueKind(*TC, ED);
```
computeAutomaticEnumValueKind() only uses TC for TypeChecker::getProtocol(). You can call ASTContext::getProtocol() instead.

CodaFi commented:
@swift-ci please clean smoke test

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1466,61 +1463,82 @@ computeAutomaticEnumValueKind(TypeChecker &TC, EnumDecl *ED) {
   llvm::SmallDenseMap<RawValueKey, RawValueSource, 8> uniqueRawValues;
 
+  auto uncheckedRawValueOf = [](EnumElementDecl *EED) -> LiteralExpr * {
+    return EED->RawValueExpr;
```
Do you really need this lambda?

CodaFi replied:
I wanted to make the raw member accesses explicit.

jrose-apple replied:
Maybe worth a comment, then?

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1466,61 +1463,82 @@ computeAutomaticEnumValueKind(TypeChecker &TC, EnumDecl *ED) {
       // autoincrement from the previous value, or start from zero if this
       // is the first element.
-      auto nextValue = getAutomaticRawValueExpr(TC, *valueKind, elt, prevValue);
+      auto nextValue = getAutomaticRawValueExpr(*TC, *valueKind, elt, prevValue);
```
You can refactor getAutomaticRawValueExpr() too:

- Get ASTContext from the Decl instead of from TC.Context
- Use ctx.Diags.diagnose() instead of TC.diagnose()

slavapestov reviewed (commented):
Can you please keep the multi-file test case around? It's still useful to test that we emit a diagnostic instead of crashing when IRGen kicks off the request but the raw values are somehow invalid.

CodaFi commented:
There's a corresponding IRGen test already for that

CodaFi commented:
Okay, integrated your feedback. Some changes from the first go around after having a chat with @jrose-apple 

- The call to `getInterfaceType()` on the enum element is *essential* to a functioning standard library being produced.  I have no idea why this is the case, but I'm investigating.  For now, it stays.
- The tests have been restored.  In particular, we're going to keep diagnosing non-primaries because you could easily imagine a nightmare world where

```
// fileA.swift
@objc enum Foo : Int {
  case a 
  case b = ""
  // ...
}
// fileB.swift
func foo() {
  switch Foo.a {
  // ...
  }
}
```

Which means that IRGen might be the first caller of `getRawValueExpr()` and potentially might see NULL and fail the assert.  I wonder if we should consider moving forcing into the `isObjC` computation, but that wouldn't solve the problem just shuffle it around.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

jrose-apple commented on lib/AST/ASTScopeCreation.cpp:
```diff
@@ -1346,8 +1346,6 @@ void EnumElementScope::expandAScopeThatDoesNotCreateANewInsertionPoint(
     scopeCreator.constructExpandAndInsertUncheckable<ParameterListScope>(
         this, pl, nullptr);
-  // might contain a closure
-  scopeCreator.addToScopeTree(decl->getRawValueExpr(), this);
```
This is suspect; invalid code might still contain a closure here. (Or if we ever allow constant expressions in enum raw values instead of just literals, but then most of this will change.)

CodaFi replied:
Spoke with @davidungar about this one.  It's always going to be a `LiteralExpr`, and we make sure that strings are never interpolated in Parse, so the invariant this was checking is prevented quasi-structurally.  That last case makes removing this not a fantastic idea - and I had only dumped it because it was one of the red herrings I found prevented the stdlib deserialization test from failing and because for now it won't actually do anything.

I'm going to change the comment to be a bit more helpful then.

jrose-apple commented on lib/AST/TypeCheckRequests.cpp:
```diff
@@ -854,3 +854,36 @@ void UnderlyingTypeRequest::cacheResult(Type value) const {
+//----------------------------------------------------------------------------//
+
+bool EnumRawValuesRequest::isCached() const {
+  return std::get<1>(getStorage()) == TypeResolutionStage::Interface;
```
Strictly speaking you could consider them both cached.

jrose-apple commented on lib/IRGen/GenEnum.cpp:
```diff
@@ -1128,7 +1128,7 @@ namespace {
-      assert(target->hasRawValueExpr()
+      assert(target->getRawValueExpr()
              && "c-compatible enum elt has no raw value?!");
       auto intExpr = cast<IntegerLiteralExpr>(target->getRawValueExpr());
```
This one looks structural, actually.

jrose-apple commented on lib/PrintAsObjC/DeclAndTypePrinter.cpp:
```diff
@@ -393,7 +393,8 @@ class DeclAndTypePrinter::Implementation
 
-      if (auto ILE = cast_or_null<IntegerLiteralExpr>(Elt->getRawValueExpr())) {
+      auto *RVE = Elt->getStructuralRawValueExpr();
+      if (auto ILE = cast_or_null<IntegerLiteralExpr>(RVE)) {
```
The "or null" here was to not print values not explicitly provided by the user, but I guess that only works when printing from source anyway. Might as well drop it, or check for `isImplicit` instead (which probably isn't preserved through serialization).

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3923,10 +3931,10 @@ void TypeChecker::validateDecl(ValueDecl *D) {
+        (void)ED->getAllElements().front()->getRawValueExpr();
     }
-
+    
```
Nitpick: Please try to avoid letting spaces slip in on blank lines.

CodaFi committed:
Switch up @objc enum printing

Now that all the raw values are computed up front, change the printer to
instead require the expression not be implicit.  Previously, a NULL
element would appear instead.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
⛵️
